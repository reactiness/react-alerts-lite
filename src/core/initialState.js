export const initialState = {
    alerts: {
        current: []
    }
}