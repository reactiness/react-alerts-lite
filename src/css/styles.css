.__rta_pos_common {
  display: flex;
  align-items: flex-end;
  flex-direction: column;
  position: fixed
}

.__rta_pos_bottom-full {
  bottom: 0px;
  left: 0px;
  width: 100%
}

.__rta_pos_top-full {
  top: 0px;
  left: 0px;
  width: 100%;
}
.__rta_pos_top-left {
    top: 0px;
    left: 0px;
    width: 380px;
}
.__rta_pos_top-right {
  top: 0px;
  right: 0px;
  width: 380px;
}
.__rta_pos_bottom-right {
  bottom: 0px;
  right: 0px;
  width: 380px;
}
.__rta_pos_bottom-left {
  bottom: 0px;
  left: 0px;
  width: 380px;
}
.__rta_pos_bottom {
  bottom: 0px;
  left: 390px;
  width: calc(100% - 780px);
}
.__rta_pos_top {
  top: 0px;
  left: 390px;
  width: calc(100% - 780px);
}
