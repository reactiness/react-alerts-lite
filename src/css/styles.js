export const containerStyle = pos => ({
  position: "fixed",
  width: "100%",
  height: "35px",
  bottom: `${pos * 40}px`,
  left: "0px",
  border: "1px grey black",
  display: "inline-block"
});

export const defaultStyle = {
  verticalAlign: "middle",
  lineHeight: "35px",
  fontFamily: "Lato,Helvetica Neue,Arial,Helvetica,sans-serif",
  width: "100%",
  height: "35px",
  color: "white"
};

export const errorStyle = {
  backgroundColor: "#f44336"
};

export const basicStyle = {
  backgroundColor: "#f1f2f2"
};

export const successStyle = {
  backgroundColor: "#4CAF50"
};

export const warningStyle = {
  backgroundColor: "#ff9800"
};

export const infoStyle = {
  backgroundColor: "#2196F3"
};
