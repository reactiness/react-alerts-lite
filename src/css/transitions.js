export const fade = {
  transition: 'opacity '
}