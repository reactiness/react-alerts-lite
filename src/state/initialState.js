export const initialState = {
  alerts: {
    current: [],
    defaultProps: {}
  }
};
