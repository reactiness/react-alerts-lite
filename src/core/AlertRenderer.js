import { Alert } from "./Alert";
import React, { Component } from "react";
import TransitionGroup from "react-transition-group/TransitionGroup";
import { containerStyle, positionStyle, themeStyles } from "../css/styles";
import { positionsArray, positionSections } from "./constants";
import { transitions as stockTransitions } from "../transitions";

export class AlertRenderer extends Component {
  constructor(props) {
    super(props);
    this.selectTransition = this.selectTransition.bind(this);
    this.selectTheme = this.selectTheme.bind(this);
  }

  selectTransition(transitionString) {
    const { transitions: customTransitions } = this.props;
    try {
      if (customTransitions) {
        const customTransition = customTransitions.find(
          t => t.name === transitionString
        );
        if (customTransition) return customTransition.transition;
      }
      if (stockTransitions[transitionString])
        return stockTransitions[transitionString];
      console.warn("Transition name not found: defaulting to fade transition");
      return stockTransitions["fade"];
    } catch (error) {
      console.warn("Transition name not found: defaulting to fade transition");
      return stockTransitions["fade"];
    }
  }

  selectTheme(themeString) {
    const { customTheme } = this.props;
    try {
      if (themeString) {
        const themeStyle = themeStyles[themeString];
        if (themeStyle) return { ...themeStyle, ...customTheme };
        console.warn("Theme not found");
      }
      return { ...customTheme };
    } catch (error) {
      console.warn("Theme not found");
      return { ...customTheme };
    }
  }

  render() {
    const { alerts, remove, theme } = this.props;
    return positionsArray.map(position => (
      <TransitionGroup
        style={{ ...positionStyle.common, ...positionStyle[position] }}
        key={`__rta_pos_${position}`}
      >
        {alerts[position].map(alert => {
          const Transition = this.selectTransition(alert.transition);
          return (
            <Transition
              duration={alert.duration}
              key={`__rta_alert_${alert.id}`}
              maxHeight={alert.maxHeight} // Little bit hacky - maybe better way than this
            >
              <Alert
                theme={this.selectTheme(theme)}
                remove={remove}
                {...alert}
              />
            </Transition>
          );
        })}
      </TransitionGroup>
    ));
  }
}
