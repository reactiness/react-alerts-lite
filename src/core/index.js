import React from "react";
import { Alert } from "./Alert";
import { store } from "./AlertProvider";
import { actions } from "../state/actions";
import { defaultAlertArgs } from "./constants";

let counter = 0;

const add = args => {
  console.log(store);
  store.dispatch(
    actions.add({
      id: counter,
      ...args
    })
  );
  counter++;
};

const remove = type => {
  type
    ? store.dispatch(actions.removeType(type))
    : store.dispatch(actions.removeAll());
};

export class Alerts {
  static push(args) {
    console.log();
    add({ ...defaultAlertArgs, ...store.state.alerts.defaultProps, ...args });
  }
  static remove(type) {
    remove(type);
  }
}
