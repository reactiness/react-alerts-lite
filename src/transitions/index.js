import { FadeSlideUp } from "./FadeSlideUp";
import { FadeSlideDown } from "./FadeSlideDown";
import { FadeSlideDownUp } from "./FadeSlideDownUp";
import { Fade } from "./Fade";
import { SlideRight } from "./SlideRight";

import { transitionConstants } from "../core/constants";

export const transitions = {
  [transitionConstants.fadeSlideDown]: FadeSlideDown,
  [transitionConstants.fadeSlideUp]: FadeSlideUp,
  [transitionConstants.fadeSlideDownUp]: FadeSlideDownUp,
  [transitionConstants.fade]: Fade,
  [transitionConstants.slideRight]: SlideRight
};
