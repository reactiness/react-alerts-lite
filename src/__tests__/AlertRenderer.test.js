import React from "react";
import { AlertRenderer } from "../core/AlertRenderer";
import {
  transitionConstants,
  positionConstants,
  themeConstants
} from "../core/constants";
import { initialState } from "../state/initialState";
import { transitions } from "../transitions";
import { themeStyles } from "../css/styles";

const alert = {
  id: 0,
  closeButton: true,
  content: "this is a test alert",
  duration: 500,
  maxHeight: "200px",
  position: "top-right",
  timeout: 3000,
  transition: "fade",
  type: "basic"
};

const sortedAlerts = {
  bottom: [
    { ...alert, id: 8, type: "basic", position: positionConstants.bottom }
  ],
  top: [
    { ...alert, id: 7, type: "basic", position: positionConstants.top },
    { ...alert, id: 9, type: "basic", position: positionConstants.top }
  ],
  "top-left": [
    { ...alert, id: 1, type: "basic", position: positionConstants.topLeft }
  ],
  "top-right": [
    alert,
    { ...alert, id: 2, type: "basic", position: positionConstants.topRight }
  ],
  "bottom-left": [
    { ...alert, id: 3, type: "basic", position: positionConstants.bottomLeft }
  ],
  "bottom-right": [
    { ...alert, id: 4, type: "basic", position: positionConstants.bottomRight }
  ],
  "bottom-full": [
    { ...alert, id: 6, type: "basic", position: positionConstants.bottomFull }
  ],
  "top-full": [
    { ...alert, id: 5, type: "basic", position: positionConstants.topFull }
  ]
};

describe("Alert Renderer", () => {
  it("should render correctly", () => {
    const wrapper = shallow(<AlertRenderer alerts={sortedAlerts} />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should select the correct transition", () => {
    const wrapper = mount(<AlertRenderer alerts={sortedAlerts} />);
    const selectTransition = wrapper.instance().selectTransition;

    let transition = selectTransition(transitionConstants.slideDown);
    expect(transition).toEqual(transitions[transitionConstants.slideDown]);
    transition = selectTransition(transitionConstants.slideUp);
    expect(transition).toEqual(transitions[transitionConstants.slideUp]);
    transition = selectTransition(transitionConstants.slideUpThrough);
    expect(transition).toEqual(transitions[transitionConstants.slideUpThrough]);
    transition = selectTransition(transitionConstants.fade);
    expect(transition).toEqual(transitions[transitionConstants.fade]);
    transition = selectTransition(transitionConstants.slideRight);
    expect(transition).toEqual(transitions[transitionConstants.slideRight]);
    transition = selectTransition(transitionConstants.slideLeft);
    expect(transition).toEqual(transitions[transitionConstants.slideLeft]);
    transition = selectTransition(transitionConstants.rotateLeft);
    expect(transition).toEqual(transitions[transitionConstants.rotateLeft]);
    transition = selectTransition(transitionConstants.rotateRight);
    expect(transition).toEqual(transitions[transitionConstants.rotateRight]);
    transition = selectTransition(transitionConstants.scale);
    expect(transition).toEqual(transitions[transitionConstants.scale]);
    transition = selectTransition(transitionConstants.scaleSlideDown);
    expect(transition).toEqual(transitions[transitionConstants.scaleSlideDown]);
    transition = selectTransition(transitionConstants.scaleSlideLeft);
    expect(transition).toEqual(transitions[transitionConstants.scaleSlideLeft]);
    transition = selectTransition(transitionConstants.scaleSlideRight);
    expect(transition).toEqual(
      transitions[transitionConstants.scaleSlideRight]
    );
    transition = selectTransition(transitionConstants.scaleSlideUp);
    expect(transition).toEqual(transitions[transitionConstants.scaleSlideUp]);
    transition = selectTransition(transitionConstants.none);
    expect(transition).toEqual(transitions[transitionConstants.none]);
  });
  it("should fallback to fade transition on error or invalid keyname", () => {
    const wrapper = mount(<AlertRenderer alerts={sortedAlerts} />);
    const selectTransition = wrapper.instance().selectTransition;
    const transition = selectTransition("invalid transition key");
    expect(transition).toEqual(transitions[transitionConstants.fade]);
  });
  it("should select the correct theme", () => {
    const wrapper = mount(
      <AlertRenderer alerts={sortedAlerts} theme={themeConstants.rounded} />
    );
    const selectTheme = wrapper.instance().selectTheme;
    let theme = selectTheme(themeConstants.rounded);
    expect(theme).toEqual(themeStyles[themeConstants.rounded]);
    theme = selectTheme(themeConstants.flatRounded);
    expect(theme).toEqual(themeStyles[themeConstants.flatRounded]);
    theme = selectTheme(themeConstants.shadowed);
    expect(theme).toEqual(themeStyles[themeConstants.shadowed]);
    theme = selectTheme(themeConstants.flat);
    expect(theme).toEqual(themeStyles[themeConstants.flat]);
    theme = selectTheme(themeConstants.bordered);
    expect(theme).toEqual(themeStyles[themeConstants.bordered]);
  });
  it("should return empty theme on error or invalid keyname", () => {
    const wrapper = mount(<AlertRenderer alerts={sortedAlerts} />);
    const selectTheme = wrapper.instance().selectTheme;
    let theme = selectTheme("invalid theme key");
    expect(theme).toEqual({});
  });
  it("should call selectTheme when providing theme", () => {
    const spy = jest.spyOn(AlertRenderer.prototype, "selectTheme");
    const wrapper = mount(
      <AlertRenderer alerts={sortedAlerts} theme={themeConstants.rounded} />
    );
    expect(spy).toHaveBeenCalledTimes(10);
  });
  it("should call selectTransition when providing alerts", () => {
    const spy = jest.spyOn(AlertRenderer.prototype, "selectTransition");
    const wrapper = mount(<AlertRenderer alerts={sortedAlerts} />);
    expect(spy).toHaveBeenCalledTimes(10);
  });
  it("should apply a custom theme when provided as a prop", () => {
    const customTheme = { border: "1px solid red" };
    const wrapper = mount(
      <AlertRenderer alerts={sortedAlerts} customTheme={customTheme} />
    );
    const selectTheme = wrapper.instance().selectTheme;
    let theme = selectTheme();
    expect(theme).toEqual(customTheme);
  });
  it("should merge a custom theme with selected theme with customTheme taking precedence", () => {
    const customTheme = { border: "1px solid red" };
    const wrapper = mount(
      <AlertRenderer alerts={sortedAlerts} customTheme={customTheme} />
    );
    const selectTheme = wrapper.instance().selectTheme;
    let theme = selectTheme(themeConstants.bordered);
    expect(theme).toEqual({
      ...themeStyles[themeConstants.bordered],
      ...customTheme
    });
  });
  it("should select custom transition when provided via props", () => {
    const customTransitions = [
      {
        name: "custom-transition",
        transition: transitions[transitionConstants.scaleSlideRight]
      },
      {
        name: "custom-transition-two",
        transition: transitions[transitionConstants.scaleSlideUp]
      }
    ];
    const wrapper = mount(<AlertRenderer alerts={sortedAlerts} transitions={customTransitions} />);
    const selectTransition = wrapper.instance().selectTransition;
    const transition = selectTransition("custom-transition-two");
    expect(transition).toEqual(transitions[transitionConstants.scaleSlideUp]);
  });
});
