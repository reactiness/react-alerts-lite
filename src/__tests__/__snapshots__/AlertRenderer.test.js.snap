// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Renderer should render correctly 1`] = `
Array [
  <TransitionGroup
    childFactory={[Function]}
    component="div"
    key="__rta_pos_bottom"
    style={
      Object {
        "alignItems": "flex-end",
        "bottom": "0px",
        "display": "flex",
        "flexDirection": "column",
        "left": "390px",
        "position": "fixed",
        "width": "calc(100% - 780px)",
      }
    }
  />,
  <TransitionGroup
    childFactory={[Function]}
    component="div"
    key="__rta_pos_top"
    style={
      Object {
        "alignItems": "flex-end",
        "display": "flex",
        "flexDirection": "column",
        "left": "390px",
        "position": "fixed",
        "top": "0px",
        "width": "calc(100% - 780px)",
      }
    }
  />,
  <TransitionGroup
    childFactory={[Function]}
    component="div"
    key="__rta_pos_top-left"
    style={
      Object {
        "alignItems": "flex-end",
        "display": "flex",
        "flexDirection": "column",
        "left": "0px",
        "position": "fixed",
        "top": "0px",
        "width": "380px",
      }
    }
  />,
  <TransitionGroup
    childFactory={[Function]}
    component="div"
    key="__rta_pos_top-right"
    style={
      Object {
        "alignItems": "flex-end",
        "display": "flex",
        "flexDirection": "column",
        "position": "fixed",
        "right": "0px",
        "top": "0px",
        "width": "380px",
      }
    }
  />,
  <TransitionGroup
    childFactory={[Function]}
    component="div"
    key="__rta_pos_bottom-left"
    style={
      Object {
        "alignItems": "flex-end",
        "bottom": "0px",
        "display": "flex",
        "flexDirection": "column",
        "left": "0px",
        "position": "fixed",
        "width": "380px",
      }
    }
  />,
  <TransitionGroup
    childFactory={[Function]}
    component="div"
    key="__rta_pos_bottom-right"
    style={
      Object {
        "alignItems": "flex-end",
        "bottom": "0px",
        "display": "flex",
        "flexDirection": "column",
        "position": "fixed",
        "right": "0px",
        "width": "380px",
      }
    }
  />,
  <TransitionGroup
    childFactory={[Function]}
    component="div"
    key="__rta_pos_bottom-full"
    style={
      Object {
        "alignItems": "flex-end",
        "bottom": "0px",
        "display": "flex",
        "flexDirection": "column",
        "left": "0px",
        "position": "fixed",
        "width": "100%",
      }
    }
  />,
  <TransitionGroup
    childFactory={[Function]}
    component="div"
    key="__rta_pos_top-full"
    style={
      Object {
        "alignItems": "flex-end",
        "display": "flex",
        "flexDirection": "column",
        "left": "0px",
        "position": "fixed",
        "top": "0px",
        "width": "100%",
      }
    }
  />,
]
`;
